### DISCUSSION TITLE
Please add support for PostgreSQL

### DISCUSSION THREAD

**Discussion by beedaddy** [wavelog/wavelog#82] [Ideas]
Since Postgresql is a very popular database system nowadays, I would like to ask for postgres support in the future. It would be really good to have a choice here, since running mysql just for one single service feels uncomfortable. I would also be happy to help with testing.

**Comment by int2001** [comment]
Uh. that's a really huge Feature-Request.
All libraries i've seen for "Multi-DB-Support" are not well written. There are always performance-issues because you've to workaround. Perhaps that changed during the last years, but i don't think so.

And a complete switch to Postgres won't be possible, because many hams are using Hosters which don't provide postgres.

you're right in one point: PGSQL is the "better Database", but it is even more complex to configure for the User.

**Comment by beedaddy** [comment]
Yes, I know it's would be a huge effort. Perhaps it could be considered again if there would ever be a major rewrite or so. At least many web application I know of do support both mysql and postgresql (Wallabag, Nextcloud, Roundcube, …). Having a choice would be really cool. As I said, maybe it's something worth to consider for the long term.

**Comment by int2001** [comment]
yes, maybe an "enterprise Version" one day. But currently not on the Roadmap.
I know Roundcube/Wallabag an so on. Difference is: less SQL-Accesses and - tbh - not very performant there. Yes, depends on machine.

**Comment by tomswartz07** [comment]
Fully understanding that the app is designed with MariaDB/MySQL in mind-

After an extremely brief look at the code, it seems that dbforge is used as the database driver- which natively has support for Postgres as well as MariaDB/MySQL- has anyone *tried* using Postgres as the backend database? If so, were there any issues?

I'll give it a go to see if there are any serious roadblocks, but it literally seems to be the case that Postgres support is built into the code: https://github.com/wavelog/wavelog/tree/master/system/database/drivers/postgre

**Comment by dl8dtl** [comment]
I'd like to second this request.

While I have containerized services like wavelog, I'd very much prefer a single (and powerful) database server rather than having a mysql here, and another DB there. This in particular eases things like maintaining regular database backups from all databases.

Besides, I guess it could substantially improve database performance on installations with a large userbase (think of DARC providing the service to its members, for example).

Alas, while I have quite some programming skills, none of them belong to PHP or JavaScript, so I cannot personally offer help except maybe in form of a donation.

**Comment by AndreasK79** [comment]
We can agree that PostgreQSL is superior. That's not where the problem lies.

Rewriting Wavelog to change the db engine is a huge effort, which we have no plans for. Remember, this is a hobby project, where we spend our free time. Our time is much better spent fixing bugs, improving existing functionality and adding new functionality to improve Wavelog.

It is already tagged "wont fix", and it will probably stay that way.

**Comment by dl8dtl** [comment]
I can fully understand your hobby project sentence – having contributed and maintained many opensource projects myself during the past decades. Some of them have taken off, and eventually found new maintainers, some are basically dead, some, I am still with.

Anyway, as tomswartz07 already mentioned, it appears the underlying database framework would already support a number of different database backends. `database.sample.php` contains:
```
|       ['dbdriver'] The database driver. e.g.: mysqli.
|                       Currently supported:
|                                cubrid, ibase, mssql, mysql, mysqli, oci8,
|                                odbc, pdo, postgre, sqlite, sqlite3, sqlsrv
```
So I guess it basically boils down to how well standard-conforming the SQL statements in the code actually are. According to my observations, PostgreSQL is sometimes quite strictly adhering to the SQL standard, where other backends might be less picky.

**Reply by int2001** [reply]
From my work- and hobby experience those "drivers" are a trap.
I've seen no "middleware" or "driver" which supports optimzing SQLs for all the engines.
And even if there's one: Would be a lot of effort to refactor everything. So as conclusio: it Isn't planned atm to support Postgres.

we moved away from the proprietary CI-Approach with their "SQL-Model" for a reason --> Performance.

**Reply by carstencodes** [reply]
I see. I have also made my experiences. Many of the PHP applications I am hosting are really on mariadb and get a huge performance boost when switching to postgres. Namely nextcloud and pixelfed. Therefore, why not give it a try?

**Reply by tomswartz07** [reply]
I don't have my test instance handy, but as memory serves- I could manually run the [`install.sql`](https://github.com/wavelog/wavelog/blob/e4d17a2a07be52b12420cb351b35b72ab4c9e8fa/install/assets/install.sql) scripts successfully within Postgres- everything provisioned correctly. However, Wavelog specifically will not start due to some hard-coded checks for Maria/MySQL. 

I began investigating the possibility of resolving those checks, but never fully completed the process to create a valid PR.

If time permits, I may revisit this work.

**Reply by int2001** [reply]
its an easy tradeoff for us with this hobby-project:
Postgres:
- Yes, beats mySQL in nearly every category
- more complex to maintain / install to the enduser

MySQL:
- widely available. 90% of the hosters are supporting it (we don't want to lose the people with simple webspaces)
- scales (hugest known installation handles over 15Mio QSOs with more than 3k Users)
- Code in Wavelog is already optimized for mySQL. (Fun-Fact: wasn't possible with the "magic-driver" Codeigniter provided.)

In General:
- No known "general DB-Driver" is able to hold it's promises when it comes to exchangeable DB-Engines AND performance/scaling.
- One need to optimize SQLs / DB-Structure for the used Engine (huge Effort, not only change backend-driver)

Conclusio:
- Add Postgres-Support: Double nearly all critical SQLs and optimize them for Postgres as well: isn't woth it, since mysql proved scaling
- Switch complety to Postgres: Lose the people who have no postgres-hoster.
- Use some magic driver: Refactor everything to a meta-SQL-Language without specific optimizations. performance will only scale with more/better hardware instead of good code.

As you can see, there's no real need to switch to Postgres compared to the effort / side-effects listed here.

So i kindly ask: **Whats the case - besides personal preferences - to have postgres instead of mysql?**
* Performance? No - proved that its working very good at scale with mySQL
* Not having multiple Databases in your homelab? - No, nearly every modern OSS-Project comes with an own preferred engine like NoSQL-DBs, mysql, redis, postgres. Thats IT
* Compatibility to most of the hosters out there? - No, nearly every hoster in this segment offers mySQL. and keep in mind: it's not an enterprise-app which needs Snowflake, OracleDB, $enterpriseDBEngine

**Reply by int2001** [reply]
> I see. I have also made my experiences. Many of the PHP applications I am hosting are really on mariadb and get a huge performance boost when switching to postgres. Namely nextcloud and pixelfed. Therefore, why not give it a try?

Reason is simple: most of those projects didn't optimize their code for Database. They simply use the DB without any tweaks in their SQLs. And if you switch to postgresql it got a one-time boost. But: When DB grows, it's essential to optimize the SQLs. Otherwise the problem is only shifted to a later time. Seen that very often.
